vllm/model_executor/layers/quantization/awq_triton.py

```diff
@@ -0,0 +1,549 @@
+                                                 split_k_iters, thx, thy)
+        print(f"Torch result:iweights_torch = {iweights_torch}")
+
+    if use_torch and use_triton:
+        diff = iweights_torch - iweights_triton
+        error = torch.sum(torch.sqrt(diff * diff))
+        print(f"error = {error}")
+
+
+def awq_gemm_triton(input: torch.Tensor, qweight: torch.Tensor,
+                    scales: torch.Tensor, qzeros: torch.Tensor,
+                    split_k_iters: int) -> torch.Tensor:
+    M, K = input.shape
+    N = qweight.shape[1] * 8
+    grid = lambda META: (
+        triton.cdiv(M, META['BLOCK_SIZE_M']) * triton.cdiv(
+            N, META['BLOCK_SIZE_N']),
+        split_k_iters,
+    )
+    awq_group_size = 128
+    block_size_m = 32
+    block_size_n = 32
+    block_size_k = 32
+
+    result = torch.zeros((M, N), dtype=torch.float16, device=input.device)
+
+    # A = input, B = qweight, C = result
+    # A = M x K, B = K x N, C = M x N
+    awq_gemm_kernel[grid](input,
+                          qweight,
+                          result,
+                          qzeros,
+                          scales,
+                          M,
+                          N,
+                          K,
+                          awq_group_size,
+                          input.stride(0),
+                          input.stride(1),
+                          qweight.stride(0),
+                          qweight.stride(1),
+                          result.stride(0),
+                          result.stride(1),
+                          qzeros.stride(0),
+                          qzeros.stride(1),
+                          scales.stride(0),
+                          scales.stride(1),
+                          BLOCK_SIZE_M=block_size_m,
+                          BLOCK_SIZE_N=block_size_n,
+                          BLOCK_SIZE_K=block_size_k,
+                          SPLIT_K=split_k_iters)
+
+    return result
+
+
+# input   - [N, K]
+# qweight - [K, M // 8]
+# qzeros  - [K // G, M // 8]
+# scales  - [K // G, M]
+# split_k_iters - parallelism along K-dimension, int, power of 2.
+def awq_gemm_torch(input: torch.Tensor, qweight: torch.Tensor,
+                   scales: torch.Tensor, qzeros: torch.Tensor,
+                   split_k_iters: int) -> torch.Tensor:
+    input_rows, input_cols = input.shape
+    qweight_rows, qweight_cols = qweight.shape
+    scales_rows, scales_cols = scales.shape
+    print(f"awq_gemm_torch:input_rows = {input_rows} input_cols = {input_cols}"
+          f" qweight_rows = {qweight_rows} qweight_cols = {qweight_cols}"
+          f" scales_rows = {scales_rows} scales_cols = {scales_cols}")
+    weights, zeros = awq_dequantize_torch(qweight, scales, qzeros,
+                                          split_k_iters, 0, 0)
+    return torch.matmul(input, weights)
+
+
+def test_gemm():
+    print("=" * 10 + " TESTING GEMM " + "=" * 10)
```

Comment:
Could you move the test code to a separate file under `vllm/tests`, to be run with pytest?

Reply:
Yes, planning on doing it, will update when done.

Reply:
Tests done